src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs

```diff
@@ -148,25 +148,32 @@ private abstract class CompilationTrackerState
             /// State used when we potentially have some information (like prior generated documents)
             /// but no compilation.
             /// </summary>
-            private sealed class NoCompilationState(CompilationTrackerGeneratorInfo generatorInfo)
-                : CompilationTrackerState(generatorInfo)
+            private sealed class NoCompilationState : CompilationTrackerState
             {
+                public NoCompilationState(CompilationTrackerGeneratorInfo generatorInfo)
+                    : base(generatorInfo)
+                {
+                    // The no compilation state can never be in the 'DocumentsAreFinal' state.
+                    Contract.ThrowIfTrue(generatorInfo.DocumentsAreFinal);
```

Comment:
note: this is a *new* contract call.  but i've written in the pr why i think it is correct.  the contract calls then help enforce these sorts of invariants.